Flaky test: test_issue_496_2 deadlocks on MS SQL (MERGE WITH (HOLDLOCK) conversion deadlock)

`test_issue_496_2` fails intermittently on MS SQL, now with a deadlock instead of the
`PRIMARY KEY` violation fixed in #747:

```
[ERROR] Tests run: 31760, Failures: 1, Errors: 0, Skipped: 0
[ERROR] MsSqlTestCase>PluggableBackendImplTestCase.test_issue_496_2:1282
        » java.lang.RuntimeException: com.microsoft.sqlserver.jdbc.SQLServerException:
          Transaction (Process ID 54) was deadlocked on lock resources with another process
          and has been chosen as the deadlock victim. Rerun the transaction.
```

```
at org.opends.server.backends.jdbc.Storage$WriteableTransactionTransactionImpl.put(Storage.java:365)
at org.opends.server.backends.jdbc.Storage$WriteableTransactionTransactionImpl.update(Storage.java:436)
at org.opends.server.backends.pluggable.PluggableBackendImplTestCase$8$1.run(PluggableBackendImplTestCase.java:1264)
at org.opends.server.backends.jdbc.Storage.write(Storage.java:184)
...
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: ... deadlock victim ...
at org.opends.server.backends.jdbc.Storage.execute(Storage.java:88)
at org.opends.server.backends.jdbc.Storage$WriteableTransactionTransactionImpl.upsert(Storage.java:397)
```

Seen in [run 30627712217, job `build-maven (ubuntu-latest, 26)`](https://github.com/OpenIdentityPlatform/OpenDJ/actions/runs/30627712217/job/91146682586)
— the only failure out of 31760 tests; every other matrix cell (JDK 11/17/21/25, macOS, Windows)
was green, so it is timing dependent.

## Analysis

`test_issue_496_2` (`PluggableBackendImplTestCase.java:1229-1288`) starts 8 threads, each running
1023 write transactions that `update()` **the same** key `"key"` of the same tree.

For MS SQL the upsert (`Storage.java:392`) is:

```sql
merge into <table> WITH (HOLDLOCK) old using (select ? h,? k,? v) new
  on (old.h=new.h and old.k=new.k)
  WHEN MATCHED THEN UPDATE SET old.v=new.v
  WHEN NOT MATCHED THEN INSERT (h,k,v) VALUES (new.h,new.k,new.v);
```

`HOLDLOCK` is `SERIALIZABLE`: the search phase of the `MERGE` takes a **shared** range lock and, unlike
under `READ COMMITTED`, holds it until the transaction ends; the `WHEN MATCHED` branch then has to
convert it to an exclusive lock. Two concurrent transactions that both hold S on the same key and
both want X deadlock on the conversion, and SQL Server kills one of them. Connections are
`autoCommit=false` / `TRANSACTION_READ_COMMITTED` (`CachedConnection.java:73-74`), so the whole
`WriteOperation` is one such transaction.

Contention is at its maximum in this test: the MS SQL DDL is `h char(128),k varbinary(max),v image,primary key(h)`
(`Storage.java:287`) — the key is a single index key — and all 8 writers hit that one row.

`Storage.write()` (`Storage.java:181-192`) rolls back and rethrows, so error 1205 propagates to the
caller even though the message explicitly asks to rerun the transaction.

So the `WITH (HOLDLOCK)` added in #747/#750 (commit b297c2fad1) removed the non-atomic-MERGE race
but replaced it with a conversion-deadlock race in the same test.

## Proposed fixes

1. Use `WITH (HOLDLOCK, UPDLOCK)` on the `MERGE` target. `UPDLOCK` makes the search phase take an
   update lock, which is incompatible with another `UPDLOCK`, so the second transaction *waits*
   instead of forming a lock cycle. This is the canonical MS SQL upsert idiom and keeps the
   atomicity that #747 needed.
2. Retry the transaction in `Storage.write()` on the retriable serialization errors — MS SQL 1205 /
   SQLState `40001`, PostgreSQL `40P01` — a bounded number of times with a small backoff. The
   transaction is already rolled back at that point, so re-running the `WriteOperation` is safe, and
   deadlocks stay possible under enough concurrency whatever locking hints are used.

## Related, but a separate defect

`WriteableTransactionTransactionImpl.update()` (`Storage.java:425-439`) issues its `read()` as a
separate statement under `READ COMMITTED`, without any lock, and only then calls `put()`. Two
concurrent transactions can read the same `oldValue`, compute from it and both write — a lost update.
Neither `HOLDLOCK` nor `UPDLOCK` on the `MERGE` closes that window, because the read happens before
it. The test does not assert on the value (it writes random UUIDs), so it does not fail on this, but
`WriteableTransaction.update()` is expected to be atomic by its callers. Worth confirming and
tracking separately.

## Not related to the CI change

The `rerunFailingTestsCount=3` removal of #801/#804 did not unmask this. On earlier runs where the
setting was still in the pom — e.g. run 30623169474 (`AliasTestCase.test_persistent_search_reports_repeated_changes`)
and run 30614325266 — a single failing `@Test` also failed the build immediately, with no `Flakes:`
section and no rerun. The TestNG provider ignored the option, so this test would have failed the same
way before that change.

--- a/opendj-server-legacy/src/main/java/org/opends/server/backends/jdbc/Storage.java
+++ b/opendj-server-legacy/src/main/java/org/opends/server/backends/jdbc/Storage.java
@@ -389,8 +389,8 @@ public class Storage implements org.opends.server.backends.pluggable.spi.Storage
 					statement.setBytes(3, value.toByteArray());
 					return (execute(statement) == 1 && statement.getUpdateCount() > 0);
 				}
-			}else if (driverName.contains("microsoft")) { //ANSI MERGE with ; WITH (HOLDLOCK) makes the upsert atomic: without it SQL Server MERGE can race two concurrent NOT MATCHED inserts of the same key into a PRIMARY KEY violation
-				try (final PreparedStatement statement = con.prepareStatement("merge into " + getTableName(treeName) + " WITH (HOLDLOCK) old using (select ? h,? k,? v) new on (old.h=new.h and old.k=new.k) WHEN MATCHED THEN UPDATE SET old.v=new.v WHEN NOT MATCHED THEN INSERT (h,k,v) VALUES (new.h,new.k,new.v);")) {
+			}else if (driverName.contains("microsoft")) { //ANSI MERGE with ; WITH (HOLDLOCK) makes the upsert atomic: without it SQL Server MERGE can race two concurrent NOT MATCHED inserts of the same key into a PRIMARY KEY violation. UPDLOCK is required on top of it: with HOLDLOCK alone the search phase takes a shared lock that the WHEN MATCHED update then has to convert to an exclusive one, so two concurrent upserts of the same key deadlock on the conversion; an update lock is taken right away and makes the second transaction wait instead
+				try (final PreparedStatement statement = con.prepareStatement("merge into " + getTableName(treeName) + " WITH (HOLDLOCK, UPDLOCK) old using (select ? h,? k,? v) new on (old.h=new.h and old.k=new.k) WHEN MATCHED THEN UPDATE SET old.v=new.v WHEN NOT MATCHED THEN INSERT (h,k,v) VALUES (new.h,new.k,new.v);")) {
 					statement.setString(1, key2hash.get(ByteBuffer.wrap(key.toByteArray())));
 					statement.setBytes(2, real2db(key.toByteArray()));
 					statement.setBytes(3, value.toByteArray());

